### DISCUSSION TITLE
Potential Roadmap Atom Community Release

### DISCUSSION THREAD

**Discussion by kaosine** [atom-ide-community/atom#393] [Ideas]
For anybody reading this wanting to help out but unsure where to look, here's all the major repos undergoing change:
 
* [Atom Core](https://github.com/pulsar-edit/pulsar)
 
* [Atom Frontend/Docs](https://github.com/kaosine/ac-website/) This is my personal repo for working on the website for this. Docs need refactoring and to get rid of custom tags while I work on the getting the front working. I should have things setup and easy to understand

* [Atom Backend/API](https://github.com/confused-Techie/atom-community-server-backend-JS) @confused-Techie 's repo, if you have any question for now ask him. More explanation below

# Major Development Items (all these are in some form of progress as of right now but all could use some help here and there)
1. CI Working Again 
  1. The current one semiworks, but this is one major sticking point as of right now. (this point will have to be updated as I get more current info). 
  1. We are currently using a older windows image in azure that is getting ready to be discontinued and just generally building the things needed for the pipelines
1. Continue upgrading Electron
   1. We're up to V12 right now thanks to a PR by Mauricio, but we're still behind farther than we should be
   1. Realistically we should probably stay at least 1-2 versions with in the current release in the future once we have  steady dev flow going
1. API backup
   1. This involves backing up the atom.io database and a few other components @confused-Techie 
   1. This is how the Windows and Mac apps auto-update and there should be some way to possibly integrate this within linux as a option? 
   1. This is also vital to some workflows like getting the auth token for github with viewing the packages, updating and modifying entries on there.
   1. Switch to apm-papm(this can be found under the repos under the organization here for atom-community) @aminya 
1. Website and documentation
   1. This involves bringing the documentation and the site back online
   1. The site needs to be a good landing point outside of the github for all the resources we have.
   1. The documentation needs to be refactored and finished (these have been pulled from the old flightmanual repo)
   1.  Update branding - knowing how Microsoft is about branding with projects it's taken over and either sunsetted or used the name for their own projects, we've decided it's best we rebrand and tentatively the name is going to be Pulsar from discussions we've had in the discord(see #polls channel for more on this)
 
I'm currently working on this last part with the website and documentation in my personal ac-website repo with hopes to push it up soon. If anyone wants to help with fixing the current documentation to allow for it to function properly and not throw errors in compilation you're more than welcome to pop over and take a look. I'm more concerned with trying to fix the layout and add more content to the front pages right now.

Not to mention this is a preliminary roadmap to be possibly converted to the projects layout. Any updates and clarifications would be helpful so I can tag who is working on what and add as necessary. These updates will likely push us to a potential 2.x release once we have the api, core, and everything else in a proper working function. Especially thanks to Mauricio's updates bringing us to electron 12 and fixing a bunch of other things once his changes are filed in. If I tag the wrong person let me know, but I'm tentatively tagging those I know who are working on specific things for the current endgoals we have at this point.

**Comment by kaosine** [comment]
Also not sure what the potential deadlines we want to set for ourselves as well. I know we have until the end of the year at the bare maximum to make sure all the API's and CI point to our new places. But It might be helpful to set something for maybe the end of July to get a dev version out at the very least?

**Reply by confused-Techie** [reply]
My goal is to have everything for the API (except Github Auth dependent features) working by the end of this month. But as soon as we decide on a rebrand we can begin looking at the Auth setup for Github

**Comment by meadowsys** [comment]
_peek_ if I may add, I may attempt to rewrite some native addons to use napi and do some monorepo-ing?

**Reply by kaosine** [reply]
I'm going to have to ask you to clarify parts of this to be honest. The mono-repoing especially because I'm not entirely sure how to word this part and I don't want it to turn out weird.

**Reply by meadowsys** [reply]
its just taking all the `language-*` packages and adding them into the main repository

**Comment by confused-Techie** [comment]
For anybody reading this wanting to help out but unsure where to look, here's all the major repos undergoing change:

* [Atom Core](https://github.com/AtomRebirth/atom)
* [Atom Frontend/Docs](https://github.com/kaosine/ac-website/)
* [Atom Backend/API](https://github.com/confused-Techie/atom-community-server-backend-JS)

**Comment by Digitalone1** [comment]
I suggest other two possible points.

1. Changing the name of the project since I read somewhere, not remember where, that atom is a mark registered and Microsoft has its rights. If it's true, we have to look for another name.
2. Changing the licence to GPL? Why MIT? I don't like someone does like Microsoft that picks Code and then adds its own without sharing with the community.

**Reply by DeeDeeG** [reply]
Re-licensing existing code is tricky, because... you have to contact all the original authors and get their permission, or remove/revert their code. For doing that properly, I'd hope for a consultant from one of the open source orgs to advise on that. linuxfoundation.org maybe?? Tricky for Atom, because there have been so many contributors.

**Reply by Digitalone1** [reply]
You're right, but the new code with changes/upgrades/fixes can be licensed as GPL. The old one would be always available as MIT, maybe archived somewhere, but the updated master branch could be GPL.

**Reply by DeeDeeG** [reply]
I keep forgetting how permissive MIT is. Okay, so as long as we maintain the MIT notice to cover the parts where that still applies, moving to GPL for the overall project should be doable. GPL just puts the limit that redistributed code needs to have the source published somewhere. It's not against anything in the MIT license to go ahead and add that additional requirement. Can be discussed by the community what is preferred in regards to license.

**Reply by Digitalone1** [reply]
Not only that, the derivative work must be distributed under the same or equivalent license terms.

**Reply by aminya** [reply]
I don't like the idea of changing the license to GPL or evilization of other companies/organizations. The limitations of GPL do not allow us to freely use code from the outer world, and many companies might ignore Atom altogether. For example, an Intel library like MKL is not compatible with GPL, which means we cannot use that library for our native modules. Or, the companies that distribute proprietary packages like Tabnine will become incompatible.

**Reply by ThatXliner** [reply]
There's LGPL... but if I'm not mistaken, that kind of just applies to libraries.

> Or, the companies that distribute proprietary packages like Tabnine will become incompatible.

What if we just make it so the editor itself will be open source, not necessarily the packages. I still would love to be able to freely edit the extensions instead of having to manually go to where the extensions are stored (e.g. VSCode)

**Reply by mauricioszabo** [reply]
I would prefer to leave the license as MIT. As much as I don't like some practices of some companies, I don't think re-licensing as GPL will be a step on the right direction or if _will even help_...

**Comment by kaosine** [comment]
Updated with the information from our renaming discussion in the discord in the #polls channel. Might be best to start over at 1.x but be clear of feature parity with Atom 1.6x. Just as a clear break from Atom and as a restart for us.

**Comment by ghost** [comment]
I'm a tech writer and use Atom for writing, so would be happy to help with the documentation.

**Reply by kaosine** [reply]
Well I do have the ac-website repo under my stuff that will eventually be pushed here but all of it probably needs updating and refactoring in the docs folder anyways. So many extraneous tags and I've had to take a break thanks to flare ups in pain. I've even got a template section I've set up and had partially finished the getting started section I think(?).

**Reply by ghost** [reply]
Ok, so if you like I could take a look? Can you give me the URL please?

**Reply by ThatXliner** [reply]
I believe @kaosine's ac-website repo is https://github.com/kaosine/ac-website and it's currently living at https://kaosine.github.io/ac-website/

**Reply by ghost** [reply]
w00t :) Will take a beak :)

**Reply by ghost** [reply]
So FYI, I'm a total n00b on collaborating on open-source projects.
I have a tendency to do some analysis first, to work through any issues I can see, then start creating issue tickets to keep track of everything.
For instance, 
* fixing the the images and hyperlinks is relatively simple using find & replace, but we need a long-term URL. This would therefore be a low-priority fix at this time
* Do you want the help totally refactored & rejigged to make it "not-atom-editor" or fix what's broken here and now and rejig it all later?

## Initial thoughts

* Remove the idea of "chapters" in a book -- I can't remember the last time I read (or wrote) a manual, and everything's online these days.   
* If anyone wants offline doco, they can just download the repo (long-term it could be worth using a site builder so users can spin up the doco locally -- e.g., Hugo, Jekyll, etc)
* End user documentation needs to address fundamental use-cases (e.g., How do I add a project, How do I add additional functionality) starting with an overview then drilling down on the detail
* Each article (or page) in the system needs to state up-front
  * what the article is explaining
  * what the user needs in place before they start making changes
  * how to make changes
  * Related information
  * A way to get more help
* End user documentation has a tendency to forget to address 6 fundamental questions:
  * Who it's for
  * What they need to know or have in place beforehand
  * When it's possible to do the "thing"
  * Where the "thing" is done (e.g., just saying SomeObscureSetting::SomethingElse isn't helpful. What's helpful is to direct the user to specific information on the setting up-front so they can look at it before they begin)
  * Why they might want to do it (or not do it if there's some kind of cost)
  * How they do the "thing"

Comments? Opinions? Objections? Leave them below :D

**Reply by icecream17** [reply]
I'm not opposed to chapters like in a book since many documentation structures are similar. (rust and python come to mind:)
https://python-chess.readthedocs.io/en/latest/
https://doc.rust-lang.org/book/

**Reply by kaosine** [reply]
Yeah the idea of a book structure honestly makes sense, since it'll probably make sense as a guide from start to finish of beginner friendly topics to the more advanced content. That's one thing I did like from the original atom documentation that's in there now that will need significant refactors and going off the outline I found in the original repo it's nowhere near done either. There's no telling how much work needs to be done to:

1. Complete it
2. Refactor for clarity
3. Get rid of extraneous stuff we probably don't need either in the "code" of the documentation or things that just are flat out not relevant.

Any help on those once I am able to start functioning again would be helpful, since writing documentation on my own isn't likely to come out exactly right even using tools like grammarly and the like to refine it(which I will be using externally and if I can figure it out I might fork out the documentation part and pull it in as a dependency for that folder(if I can figure out how to embed it in that way?). That way it can also be loaded into AC/Pulsar itself as needed. But most of that is way down the road. Just getting it functional looking and clear to anyone reading is the first goal first and foremost along with just completing as much of it as we can too.

**Reply by ghost** [reply]
Okay, so what we're missing here is the answer to the first question I posed:

> Who is the content for?

The answer helps determine the combination of:

* Delivery Mechanism -- the system the information is delivered (e.g., CMS, site builder, etc)
* Delivery structure -- the way the information is conveyed (e.g., tutorial/manual, knowledgebase/help, etc)

For example, Atom Editor has multiple audiences:

1. Those who want to use the editor as a tool (e.g., writers, tech writers, developers, engineers)
2. Those who want to develop additional functionality for the system (e.g., developers, engineers, architects etc)

## Use as a tool

The people who want to use the system as a tool don't need a tutorial/manual. They need the basics to do their jobs, which includes:
* how to add git projects
* how to change theme/appearance
* how to add additional functionality with add-ons/plugins/packages, etc
* how to commit and push their content to the repository.

In this case, a knowledgebase/help system is more appropriate

## Enhancements to the system

Those who want to add functionality to the system need to know:
* the underlying architecture, language, etc
* the package system/processes to add new functionality
* the process for working with the team to add enhancements, testing, PRs, etc
* API definitions, syntax, examples, etc

For this audience you can go with a "manual/tutorial" style help system and/or a knowledgebase.

**Reply by confused-Techie** [reply]
I really like how @ljsinclair has described the content, and think it would be really cool to really have two sections for the documentation. The way I see it, would be best shown as, 
the user clicking onto the docs page nav, then being shown the two options
* Quick Start 
* Documentation

This would naturally cause each group of documentation consumers to go the proper route, since end users, wanting to use it as a tool would click the 'Quick Start' and that can be tutorials, then those wanting to enhance the system would go towards the documentation.

This could be a simplistic way to create that divide in the docs a bit more naturally.

**Reply by ghost** [reply]
Some additional comments

* Glossaries are your friend -- any term, standard, acronym, jargon goes into the glossary, ideally with a hyperlink to an overview page for further information
* FAQs are a subset of a glossary, but with questions and are equally useful BUT only when there are articles with direct answers to the questions -- don't create endless scroll articles and expect the user to work out what to do next.
* Content duplication is bad and causes all sorts of maintenance issues. Ideally the system will have the concept of the "include" file (this is Jekyll's term for it), but in short, it's reusable content.
* It's ok to start off with broad concepts, but break detail down into child articles.
* It's bad form in documentation *and* software development to create endless scroll files. If you don't write a programme in one huge file (prior to compiling and deployment) then it shouldn't happen in doco (I'm looking at you, [VScode](https://code.visualstudio.com/docs/editor/editingevolved)

**Comment by ghost** [comment]
If you don't change the link to Pulsar Edit, you'll see that AtomRebirth isn't being worked on.
